episodes/12-fire-forget-tasks.md

```diff
@@ -0,0 +1,392 @@
+- "How do execute work in parallel?"
+::::::::::::::::::::::::::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::: objectives
+- "First objective."
+::::::::::::::::::::::::::::::::::::::::::::::::
+
+A Chapel program always start as a single main thread. You can then start concurrent tasks with the `begin`
+statement. A task spawned by the `begin` statement will run in a different thread while the main thread
+continues its normal execution. Consider the following example:
+
+```chpl
+var x=0;
+
+writeln("This is the main thread starting first task");
+begin
+{
+  var c=0;
+  while c<100
+  {
+    c+=1;
+    writeln('thread 1: ',x+c);
+  }
+}
+
+writeln("This is the main thread starting second task");
+begin
+{
+  var c=0;
+  while c<100
+  {
+    c+=1;
+    writeln('thread 2: ',x+c);
+  }
+}
+
+writeln('this is main thread, I am done...');
+```
+
+```bash
+>> chpl begin_example.chpl -o begin_example
+>> ./begin_example
+```
+
+```output
+This is the main thread starting first task
+This is the main thread starting second task
+this is main thread, I am done...
+thread 2: 1
+thread 2: 2
+thread 2: 3
+thread 2: 4
+thread 2: 5
+thread 2: 6
+thread 2: 7
+thread 2: 8
+thread 2: 9
+thread 2: 10
+thread 2: 11
+thread 2: 12
+thread 2: 13
+thread 1: 1
+thread 2: 14
+thread 1: 2
+thread 2: 15
+...
+thread 2: 99
+thread 1: 97
+thread 2: 100
+thread 1: 98
+thread 1: 99
+thread 1: 100
+```
+
+As you can see the order of the output is not what we would expected, and actually it is completely
+unpredictable. This is a well known effect of concurrent tasks accessing the same shared resource at the same
+time (in this case the screen); the system decides in which order the tasks could write to the screen.
+
+::::::::::::::::::::::::::::::::::::::: discussion
+
+## Discussion
+
+- What would happen if in the last code we declare `c` in the main thread?
+- What would happen if we try to modify the value of `x` inside a begin statement?
+
+Discuss your observations.
+
+:::::::::::::::::::::::::::::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::: callout
+
+All variables have a **_scope_** in which they can be used. The variables declared inside a concurrent tasks
+are accessible only by the task. The variables declared in the main task can be read everywhere, but Chapel
+won't allow two concurrent tasks to try to modify them.
+
+::::::::::::::::::::::::::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::::: discussion
+
+## Try this ...
+
+Are the concurrent tasks, spawned by the last code, running truly in parallel?
+
+The answer is: it depends on the number of cores available to your job. To verify this, let's modify the code
+to get the tasks into an infinite loop.
+
+```chpl
+begin
+{
+  var c=0;
+  while c>-1
+  {
+       c+=1;
+      //writeln('thread 1: ',x+c);
+   }
+}
+```
+
+Now submit your job asking for different amount of resources, and use system tools such as `top`or `ps` to
```

Comment:
Include shell commands to show the Instructor and Learners how to launch with different core (and node?) counts.

Reply:
I included the shell commands to check for CPU usage. I think doing this on their current setup is good for learning, but asking them to resubmit the job with a different number of CPUs might be confusing, as there are quite a few steps, and students running Chapel on their own computers will not be able to this. I removed the sentence asking them to resubmit their job. I also included a refresher at the top of this chapter on submitting jobs with salloc. Resolved through a commit.